### ISSUE TITLE
Stack trace from MonitorObserver thread

### ISSUE THREAD

**Issue by penguinland** [pyudev/pyudev#194]
I'm using pyudev version 0.20 in a Python3 process. Sometimes, everything works fine, but sometimes the `MonitorObserver` thread crashes with this stack trace:

```
Traceback (most recent call last):
  File "/usr/lib/python3.4/threading.py", line 920, in _bootstrap_inner
    self.run()
  File "/usr/local/lib/python3.4/dist-packages/pyudev/monitor.py", line 530, in run
    for device in iter(read_device, None):
  File "/usr/local/lib/python3.4/dist-packages/pyudev/_util.py", line 167, in eintr_retry_call
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.4/dist-packages/pyudev/monitor.py", line 357, in poll
    if eintr_retry_call(Poll.for_events((self, 'r')).poll, timeout):
  File "/usr/local/lib/python3.4/dist-packages/pyudev/_util.py", line 167, in eintr_retry_call
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.4/dist-packages/pyudev/os.py", line 213, in poll
    return list(self._parse_events(eintr_retry_call(self._notifier.poll, timeout)))
  File "/usr/local/lib/python3.4/dist-packages/pyudev/os.py", line 228, in _parse_events
    raise IOError('Error while polling fd: {0!r}'.format(fd))
OSError: Error while polling fd: 8
```

I've tried looking into this, and found that file descriptor 8 is the udev `Monitor` object itself, and its associated `event_mask` is 9 (which is `select.POLLIN | select.POLLERR`). If I modify the `_parse_events` function to just log a warning and then skip the file instead of raising an `IOError`, I get that same "file" over and over, and no longer receive updates about which devices are connected and disconnected.

What can I do differently to avoid this crash (or to continue getting updates when devices are connected/disconnected despite this crash)? Thanks in advance for the help.

Not sure if relevant: I've got a custom circuit board with a whole bunch of USB switches on it, and I'm using pyudev to do things when these switches turn on and off. When this board is reset, it's common to have a couple dozen USB devices all connect or disconnect in unison, and I think that's when these stack traces happen.

**Comment by penguinland** [comment]
Oh! Also, I'm running Linux (Ubuntu 14.04), if that makes a difference.

**Comment by mulkieran** [comment]
Thanks for the report.

I suspect that you're correct and it is something about the self-referential nature of the behavior that is causing the problem.

It would be ideal if we could get to the bottom of precisely why there is an error. In pursuit of that goal, it might be reasonable to come up with a simple C app, which uses libudev to monitor udev events. Would you be able to run such a thing on your system, if provided?

Also, if this happens again, could you supply the journal contents for that time?  Something like:
```journalctl --output=verbose --since=today```
might work.

(Note that it is not absolutely certain that raising an exception on POLERR _is_ the correct thing to do. It is reasonable to believe that moving on to the next event is a better behavior. This holds true for POLLINVAL as well. It is also true that such errors should not be entirely ignored. Pyudev itself does not log, only interacts with libudev's own logging facilities.)

**Comment by mulkieran** [comment]
It's the raising exceptions, though, that keeps the interface clean. If the exception is not raised and ```_parse_events()``` simply returns, then Poll.poll() could evaluate to an empty list, which is interpreted as a timeout, incorrectly.

**Comment by penguinland** [comment]
Thanks for taking a look!

I agree that raising the exception is probably the right design, so long as something else is there to catch and handle it so the whole thread doesn't just crash.

I'd be happy to run a C app to investigate things further, though I doubt I know enough about udev to write it myself. I'll try to get you some output from `journalctl` later today.

**Comment by penguinland** [comment]
I'm running Ubuntu 14.04, so [systemd is not used](http://askubuntu.com/questions/617290/how-to-read-systemd-journal-file-on-system-that-has-no-systemd). This in turn means that `journalctl` is not available. Unfortunately, I can't upgrade the OS to something more recent. Is there a different way I can provide you with the info you want?

**Comment by mulkieran** [comment]
There isn't any self-referentiality, but there is a bit of recursive behavior.

When the MonitorObserver is running it polls for two things: read events on the Monitor and a stop event.
If it detects a read event on the Monitor, it starts the Monitor's poll function, which again polls for read events on itself. If it find any events on itself, it gets a device, or None if there is no device. It is not clear if this recursive behavior serves a purpose or not. It might make just as much sense to receive the device without polling again, who can say?

**Comment by mulkieran** [comment]
Given absence of systemd, it may not be worth tracking down the root cause, as the bug would be too old from systemd's point of view to be worth taking an interest in. However, it is still probably worth taking an interest in the Poll class and seeing if it could be made a little more robust. I think that Poll.poll() should yield a list containing every event that is made available to it.

**Comment by mulkieran** [comment]
Could you describe perhaps by a patch or a PR, the exact change you made, and the exact results?

I have an idea of a slightly more robust approach, but if there were continuously an error in polling the monitor, i.e., if it was not a transient phenomenon, no more devices would be received after the error had started to appear.

That you keep receiving the same device over and over is kind of surprising, so I'ld be interested to see how you got to that situation.

**Comment by penguinland** [comment]
The `Monitor` object's implementation of `fileno` calls `udev_monitor_get_fd` (implemented in `libudev-monitor.c`), which returns a socket stored in a struct. I don't know enough about sockets to know why `POLLERR` would be set for one, but I imagine that once that happens, the socket needs to be closed and a fresh one needs to be opened before new updates will arrive.

I think my use of pyudev is pretty vanilla:

```python
        self._context = pyudev.Context()
        monitor = pyudev.Monitor.from_netlink(self._context)
        # Note for mulkieran: self._device_event acquires self._lock before doing anything.
        observer = pyudev.MonitorObserver(monitor, callback=self._device_event)
        # We need to initialize the status of each device, and then keep the
        # status updated as devices disconnect and reconnect. So we don't miss
        # an update, we start the observer before initialization. To prevent the
        # updates from sneaking in while we're still initializing, we acquire
        # the lock first, and don't release it until we're done with
        # initialization.
        with self._lock:
            observer.start()
            for udev_device in self._context.list_devices():
                self._device_event(udev_device)
```

I don't know enough about udev to tell if this is a bad idea, but would it work to create a new `Monitor` device when we encounter this error, and just swap the new one into the `MonitorObserver`?

**Comment by penguinland** [comment]
Sure, I can whip up changes I made again, and show the results. Give me a little while...

**Comment by mulkieran** [comment]
Thanks. I think I misunderstood you a bit and that actually does make sense. I think that the code is just running around in the outermost for-loop in Monitor.Observer.run(). It hits the inner loop, but doesn't enter the body, because the device is None, so it goes back up to the outer for loop. etc.

I think you may be on to something with the idea of creating a new monitor to observe....

**Comment by mulkieran** [comment]
What happens if instead of continuing after you encounter the error, you pass? If you do that, then the list of events will be non-empty, and libudev will attempt to read the device. The interesting question is, what happens then, i.e., can the device be read?

**Comment by penguinland** [comment]
Sorry, I had a busy day and didn't get to this. I'll try it out Monday. I imagine that it can't be read because the POLLERR bit is set.

**Comment by penguinland** [comment]
I tried just printing things out when we got a POLLERR and then going through the rest of the loop (and yielding a 'r' and stuff). Here's what I found:

- _All_ events go through file descriptor 8; if that gets hosed, no updates will ever come.
- Once the POLLERR is set, no updates ever come.

Here's the change I made to the bottom of `_os/poll.py`:

```python
        for fd, event_mask in events:
            if self._has_event(event_mask, select.POLLNVAL):
                #raise IOError('File descriptor not open: {0!r}'.format(fd))
                print("penguinland had a POLLNVAL on file {} with events {}"
                        .format(fd, event_mask))
            elif self._has_event(event_mask, select.POLLERR):
                #raise IOError('Error while polling fd: {0!r}'.format(fd))
                print("penguinland had a POLLERR on file {} with events {}"
                        .format(fd, event_mask))
            print("penguinland had a new event on file {} with {}"
                    .format(fd, event_mask))

            if self._has_event(event_mask, select.POLLIN):
                yield fd, 'r'
            ...
```

and the output:

```
penguinland says the fd is 8
penguinland had a new event on file 8 with 1
penguinland had a new event on file 8 with 1
penguinland had a new event on file 8 with 1
...
penguinland had a new event on file 8 with 1
penguinland had a new event on file 8 with 1
penguinland had a new event on file 8 with 1
penguinland had a POLLERR on file 8 with events 9
penguinland had a new event on file 8 with 9
```

I definitely connected and disconnected devices after that POLLERR, but pyudev didn't notice any changes after that.

**Comment by mulkieran** [comment]
Thanks again!

**Comment by mulkieran** [comment]
The MonitorObserver is set up in a way that allows the Monitor to be shared among multiple MonitorObservers. This means, that if the MonitorObserver manages the Monitor by making a new one, and there are multiple observers, the eventual outcome could be surprising.

I'm thinking of adding a new class, MonitorMaster, that keeps its Monitor object internal (and thus can manage it w/out worrying about other observers). What do you think?

**Comment by penguinland** [comment]
I'm no longer convinced that my suggestion was a good one. If we make a new Monitor whenever the old one has trouble, we might miss updates that happen after the old one dies and before the new one is ready. We could instead always have two Monitors and try to make sure that we get any update mentioned by either one, hoping that at most one at a time will die and we can replace it before the other one dies, too. but I have no idea if that's a good assumption, and coordinating between the two Monitors sounds annoying. Perhaps the real fix is in udev itself: figure out what causes the socket's POLLERR flag to get set and avoid that entirely, rather than treating the symptoms here in pyudev.

Then again, you know way more about udev and pyudev than I do, and I defer to your expertise.

**Comment by mulkieran** [comment]
You are quite right that it would be good to find out what is causing POLLERR to be set. Nowadays, though, libudev is a part of systemd and it would be necessary to show this problem occurring in a recent version of systemd (containing a recent version of libudev).

However, I think finding a way to make pyudev more robust is a useful end in itself.

Your point about losing udev events in the interval between the error and opening a new Monitor is certainly valid.

Every event has a kernel sequence number associated with it, which could be helpful, but the events aren't necessarily processed in order of their sequence number.

**Comment by penguinland** [comment]
You make very good points. 

Tell me more about kernel sequence numbers. Is there a way of looking them up after-the-fact in logs somewhere? Perhaps when this problem happens, we could start up the new Monitor and then backfill any events that happened while the Monitors were out, based on the logs.

If that's not an option, could the MonitorMaster have some sort of callback that gets run whenever it swaps out a Monitor? That way, we can reset certain parts of our state and iterate over all devices to get the correct settings again when the new Monitor comes up, similar to the [code I showed you previously](https://github.com/pyudev/pyudev/issues/194#issuecomment-232713879).

**Comment by mulkieran** [comment]
(Some SO notes about POLLERR: http://stackoverflow.com/questions/24791625/linux-socket-handling-revents-pollerr-pollhup-pollnval.)

**Comment by mulkieran** [comment]
@penguinland I don't think option 1 could ever be really relied on, which makes option 2 seem better.

**Comment by mulkieran** [comment]
Could you find out if "udevadm monitor --udev" also stops detecting udev events when MonitorObserver hits the exception (assuming you have udevadm available)?

**Comment by penguinland** [comment]
`udevadm monitor --udev` was able to continue detecting events. Steps I took:

- started `udevadm monitor --udev` in one terminal
- started and stopped my main program several times, until I got the stack trace. When I got the stack trace, I kept my program running, though it reported incorrect status for which devices were plugged in.
- unplugged and re-plugged several USB devices. My program did not notice any changes, but udevadm printed stuff out with no trouble.
- I started and stopped my main program several more times, and got the stack trace again, but udevadm was fine the whole time.

I know so little about udev stuff that this could be totally wrong, but might there be some sort of timeout or race condition where if my code is too slow it messes up the socket? `udevadm monitor` seems very fast, and it wasn't having any problems.

**Comment by mulkieran** [comment]
Thanks, again. Yes, your explanation is possible.

This is an orthogonal issue, but I wonder if you need to list every device and monitor every event? If you are only interested in a subset of devices and events, you could filter on both list_devices() and the Monitor. If this were possible, it _might_ reduce the incidence of the stack trace a good deal.

**Comment by mulkieran** [comment]
@penguinland This is another shot in the dark, but it might be worth a try to use the ```Monitor.set_receive_buffer_size()``` method to increase the size of the buffer.

**Comment by mulkieran** [comment]
@penguinland  Also, could you try running this extremely simple test script in the same way that you ran ```udev monitor --udev``` and see how its behavior compares?

[test.py.txt](https://github.com/pyudev/pyudev/files/383872/test.py.txt)

**Comment by penguinland** [comment]
You're right that we don't need every event; we're really just looking for non-hub USB devices. I'll look into how to filter for that sort of thing. Some devices are identified by their vendorID and productID, while some are identified by the subsystem and kernel number (we did this because we have multiple devices with identical vendor and product IDs). Perhaps using two Monitors, each one filtering for one of these groups, would do the trick.

Your test.py, like `udevadm`, seems to work reliably. I started it, then started and stopped my main program several times until I got the stack trace twice. Your program stayed up the whole time.

The [documentation for `set_receive_buffer_size`](http://pyudev.readthedocs.io/en/latest/api/pyudev.html#pyudev.Monitor.set_receive_buffer_size) says "The CAP_NET_ADMIN capability must be contained in the effective capability set of the caller for this method to succeed. Otherwise EnvironmentError will be raised, with errno set to EPERM. Unprivileged processes typically lack this capability." and indeed, I lack this ability:

```
$ python3
Python 3.4.3 (default, Oct 14 2015, 20:28:29) 
[GCC 4.8.4] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import prctl
>>> prctl.cap_effective.net_admin
False
```

So, I don't think `Monitor.set_receive_buffer_size` is an option.

**Comment by mulkieran** [comment]
@penguinland Be aware that you can invoke the ```filter_by()``` method on a Monitor multiple times and the effect is a _disjunction_ of the different filters. So, it might be possible to get events for all the devices of interest on a single Monitor.

This will only have a chance be helpful if you can filter out by this means a significant number of the events caused by, e.g., the insertion of a USB device, since I imagine that this activity is the main thing that is going on while your program is running.

**Comment by mulkieran** [comment]
@penguinland If you can give this script a try in the same way as the previous one I'ld appreciate it.

[test2.py.txt](https://github.com/pyudev/pyudev/files/384596/test2.py.txt)

**Comment by penguinland** [comment]
The MonitorObserver runs as a separate thread, so sleeping in the main thread shouldn't affect anything. Did you mean for the sleep to be in the callback?

**Comment by penguinland** [comment]
...oh, I suppose you're just checking if the problem arises from a very basic MonitorObserver instead of from polling the Monitor directly like the previous test, and we can figure out whether the problem is something unusual in the callback later.

I'll go run this code now...

**Comment by mulkieran** [comment]
Exactly!

**Comment by penguinland** [comment]
Aha! test2.py worked just great, and did not crash. However, I made a test3.py which was identical except with this callback:

```python
def callback(device):
    print('starting {0.action} on {0.device_path} at {0.sequence_number}'.format(device))
    time.sleep(1)
    print('stopping {0.action} on {0.device_path} at {0.sequence_number}'.format(device))
```

and it crashes very reliably. Not the first time the callback runs, but something around the eighth time, give or take. I thought maybe it could get all the way through the first call to [_parse_events](https://github.com/pyudev/pyudev/blob/develop-0.22/src/pyudev/_os/poll.py#L99) but crashed on the second one, so I put a print statement at the top of that function. However, it crashes on roughly the fifth time into `_parse_events`. I don't see a pattern to that yet.

I'll see what I can do to speed up my callback; it looks like that'll help a bunch.

**Comment by penguinland** [comment]
I tried calling `monitor.filter_by_tag(u"idVendor")` to reduce the number of devices we're monitoring, per your suggestion (since all the devices I care about show up in lsusb and thus have vendor IDs). It no longer crashes, but it skips over certain devices that don't have a vendorID themselves but do have an ancestor device with one. Our callback walks the hierarchy of parent devices and makes a big map of all the attributes in all of them, which is why this worked in the unfiltered version (and we did that because some of the devices we identify by the KERNELS attribute, which deals with the kernel number of any ancestor device, rather than the current device itself). 

If I could filter for all devices that have an ancestor with a vendorID, that would be perfect. I suspect there are other ways to approach this, though, and I'll keep digging.

Thanks for all your help so far!

**Comment by penguinland** [comment]
What if each time the MonitorObserver called the callback, it did so by spawning it in a separate thread? That way, the MonitorObserver's polling loop could be nice and fast, but this might break assumptions made by other pyudev users regarding thread safety.

...and I suppose that if the callback is very slow and devices are plugged and unplugged a whole lot, this could act like a thread bomb. It's probably a bad idea after all.

**Comment by mulkieran** [comment]
Yes, I think the only way to cope with this would be a deque of events and a consumer thread. Either the pyudev client would have to write a callback that simply moved the events to a separate deque and returned, or pyudev would perhaps have to take on that job.

The good/bad news is that I can get this stack trace on my system by wiggling just one USB key in and out. I can do it when I slow down the MonitorObserver via a time-consuming callback or when I slow down the Monitor itself via a time-consuming loop-body.

The next step is to get this to happen in a C program, by much the same technique, so I'm giving that a try.

**Comment by mulkieran** [comment]
Here's a C program that just connects to the monitor and reads from it, with a delay of 1 second on each read.

Would you compile and run it in the usual way? My experience was that I could not reproduce the problem (or something like) it with this program. I'ld like to know if your experience is the same.

[test.c.txt](https://github.com/pyudev/pyudev/files/386087/test.c.txt)

**Comment by penguinland** [comment]
I had the same experience: this program did not crash, even though I got my main program's MonitorObserver to crash several times while it was running.

Compilation details: I installed the latest version of the libudev-dev package (204-5ubuntu20.19), then compiled with `gcc test.c -ludev`

**Comment by mulkieran** [comment]
So, the attached reads from the monitor even if a POLLERR was reported. Reading from the monitor resets it, and since I'm triggering this with all one USB key, the socket more or recovers and events can continue to be read from it. In your case, once PROCERR is encountered, it might persist, as you have many more events than I can generate. But in any case, we have to assume that an event or two was dropped.

From your point of view, right now, the thing to do might be to have  an internal deque in your app which the callback simply writes the event to, and then, again in your app, read from the deque and process the events.

I am of two minds about whether pyudev should have an internal deque or not.

[test1.c.txt](https://github.com/pyudev/pyudev/files/386975/test1.c.txt)

**Comment by penguinland** [comment]
I can get test.c to experience the POLLERRs easily, and yet it continues on just fine afterwards! Running my main program several times without restarting test.c in the middle gets it to experience several POLLERRs, but it just keeps going. Making pyudev do that would be great.

Putting the relevant info in a deque to be examined in a separate thread is normally an excellent idea. However, I'm pretty sure that the slow part of my code is walking the device hierarchy and examining the tags of the ancestor devices, and that would all have to come before adding this info to the deque, which wouldn't actually help (disclaimer: I haven't actually profiled the code; this is just an educated guess of which part is slow). I assume that a pyudev.Device object has its state change when the device it represents disconnects, and so I can't just put the whole Device in the deque to look at later. If that's not the case, and Devices are immutable, this would be a great solution.

I'm still looking into using `Monitor.filter_by`. I had started with `Monitor.filter_by_tag`, but ran into unexpected problems. Now I'm just focusing on `filter_by`, and am hopeful it'll work later today.

**Comment by mulkieran** [comment]
It is actually possible to look up data in a Device that has been removed; that is why I had thought of having a strict mode that checked if the Device still existed before returning data about it. libudev does some sort of caching which is not well documented or clearly defined, but has certainly been observed: see #142 .

I have yet to determine whether events are lost when the POLLERR occurs.

**Comment by mulkieran** [comment]
@penguinland So, as of yet, it looks like the existance of a POLLERR does not _guarantee_ that events are lost.

I ran the below two files and found that sleep.c, yielded:

```
...
seqnum: 7950
seqnum: 7951
seqnum: 7952
seqnum: 7953
seqnum: 7954
seqnum: 7955
seqnum: 7956
seqnum: 7957
seqnum: 7958
seqnum: 7959
seqnum: 7960
POLLERR
Surprising null device.
seqnum: 7961
seqnum: 7962
seqnum: 7963
seqnum: 7964
seqnum: 7965
seqnum: 7966
seqnum: 7967
seqnum: 7968
seqnum: 7969
...

```

Note that no number is missing from the list of sequence numbers. This should certainly mean that no events were lost. (If a number was missing, there might be another explanation than a dropped event, since these numbers are generated by the kernel, and don't necessarily all make it to the udev event queue.)

[no-sleep.c.txt](https://github.com/pyudev/pyudev/files/390569/no-sleep.c.txt)
[sleep.c.txt](https://github.com/pyudev/pyudev/files/390572/sleep.c.txt)

Do you find any evidence of dropped events if you try this? The no-sleep.c program is equivalent, just it doesn't sleep and should never drop an events, so its output can be used as a kind of reference.

**Comment by mulkieran** [comment]
So, I ran both these programs, and watched the output of sleep until I observed POLLERR. Then I stopped generating uevents. Then I waited until all uevents had been processed, and I observed that sleep's output was missing 16 uevents. So, yes, uevents are being dropped, and they could easily be just as important as any other uevents.

**Comment by penguinland** [comment]
It sounds like you figured out what you wanted from those two programs, so you're not waiting on me to run them. I've gotten the `Monitor.filter_by` stuff to work, and tentatively think that's enough to avoid the crashes for my application (and I'll be more confident of that next week). 

I'm still happy to help however I can to fix this for the rest of pyudev, but I'm no longer waiting on a change.

**Comment by mulkieran** [comment]
Ok. Let me know how that goes. This is a tricky problem for pyudev to solve in a general way, so I plan to approach it cautiously. PR #215 works, in the sense that the user can specify a number of retries to the MonitorObserver, but I'm quite sure that events are being dropped just as with my C test programs. So, I'm really not sure how useful it is.

**Comment by penguinland** [comment]
I'm now confident that my code is working reliably. Thanks for suggesting the filtering, and for helping figure out what's going wrong! Let me know if there's more I can do to help you fix this for everyone else.

**Comment by mulkieran** [comment]
Good to hear. I'm still considering what might be best for pyudev itself...

**Comment by mulkieran** [comment]
I'm still a bit in favor of a queue. It is very hard to observe the socket filling up, but a queue could be used for profiling and detecting that events were arriving much faster than they were being processed. But, it would be nice to be able to solve this generically.

**Comment by penguinland** [comment]
Ohh, the problem is that the socket is filling up? That makes a lot of sense, and I hadn't put that together before. I agree that a queue is a great way to deal with this.

**Comment by mulkieran** [comment]
```/proc/net/netlink``` should indicate if any drops have occurred if consulted while process is running.

**Comment by iTagir** [comment]
Hello, I have this issue when processing multiple devices in udev. I wonder if there was any solution to this issue other than putting devices to queue, which I'm already doing (does not help :( ) and filtering which also does not help in my case.

**Comment by ogayot** [comment]
Hello,

Sorry to dig up this old topic. I'm happy to open a new issue if needed.

I've been experiencing a similar problem in the ubuntu server installer, namely in a situation where a burst of udev block events occurs.

I also found a bug report [1] in the radvd project (with a fix merged) ; where the original poster mentions that:

> In case of receive buffer is filled completely up
> Netlink may use combination of POLLER and POLLIN to indicate this.
>
> In such situation this issue can only be resolved by reading from the buffer.

I tested again after applying the following patch, and it seems to have addressed the problems I had:

```diff
diff --git a/src/pyudev/_os/poll.py b/src/pyudev/_os/poll.py
index c9da7fa..bdc2880 100644
--- a/src/pyudev/_os/poll.py
+++ b/src/pyudev/_os/poll.py
@@ -107,7 +107,8 @@ class Poll(object):
             if self._has_event(event_mask, select.POLLNVAL):
                 raise IOError('File descriptor not open: {0!r}'.format(fd))
             elif self._has_event(event_mask, select.POLLERR):
-                raise IOError('Error while polling fd: {0!r}'.format(fd))
+                if not self._has_event(event_mask, select.POLLIN):
+                    raise IOError('Error while polling fd: {0!r}'.format(fd))
 
             if self._has_event(event_mask, select.POLLIN):
                 yield fd, 'r'
```

Does it sound like a sensible thing to do in pyudev? Should we expect other types of FD to be passed to the _parse_events function?

I'm happy to work on providing simple yet reproducible use-case if needed.

[1] https://github.com/radvd-project/radvd/pull/160

Best regards,
Olivier

**Comment by jmencak** [comment]
Hello, do we have a reliable reproducer for this?  We're seeing this intermittently on server starts, but so far I've failed to reproduce it by triggering udev block events.

**Comment by yarda** [comment]
> I tested again after applying the following patch, and it seems to have addressed the problems I had:
> 
```
diff --git a/src/pyudev/_os/poll.py b/src/pyudev/_os/poll.py
index c9da7fa..bdc2880 100644
--- a/src/pyudev/_os/poll.py
+++ b/src/pyudev/_os/poll.py
@@ -107,7 +107,8 @@ class Poll(object):
             if self._has_event(event_mask, select.POLLNVAL):
                 raise IOError('File descriptor not open: {0!r}'.format(fd))
             elif self._has_event(event_mask, select.POLLERR):
-                raise IOError('Error while polling fd: {0!r}'.format(fd))
+                if not self._has_event(event_mask, select.POLLIN):
+                    raise IOError('Error while polling fd: {0!r}'.format(fd))
 
             if self._has_event(event_mask, select.POLLIN):
                 yield fd, 'r'
```
> Does it sound like a sensible thing to do in pyudev? Should we expect other types of FD to be passed to the _parse_events function?
> 
We hit this problem in TuneD project as well. I can confirm this patch worked for us, but it didn't resolve the underlying problem with the small buffer, so it later also fails with:
```
  File "/usr/lib/python3.13/site-packages/pyudev/monitor.py", line 355, in poll
    return self._receive_device()
           ~~~~~~~~~~~~~~~~~~~~^^
  File "/usr/lib/python3.13/site-packages/pyudev/monitor.py", line 294, in _receive_device
    device_p = self._libudev.udev_monitor_receive_device(self)
  File "/usr/lib/python3.13/site-packages/pyudev/_ctypeslib/_errorcheckers.py", line 99, in check_errno_on_null_pointer_return
    raise exception_from_errno(errnum)
OSError: [Errno 105] No buffer space available
```
But this error is much more legible.

**Comment by yarda** [comment]
> Hello, do we have a reliable reproducer for this? We're seeing this intermittently on server starts, but so far I've failed to reproduce it by triggering udev block events.

Reproducer is simple, just lower the udev buffer size e.g. to 1 kB (or even lower) from the default 8 kB with the (`set_receive_buffer_size(1024)`) and generate some udev events.
